File: internal/sdk/typescript/src/jobs.ts
import { createHash } from "node:crypto";
import {
	clone,
	create,
	type DescMessage,
	type MessageInitShape,
	type MessageShape,
	toBinary,
} from "@bufbuild/protobuf";

import {
	CancelJobRequestSchema,
	GetJobRequestSchema,
	ListJobsRequestSchema,
	type ListJobsResponse,
	ListJobsResponseSchema,
	RequestJobRequestSchema,
} from "../../../../protocols/generated/typescript/internal/job/v1/job_service_pb.js";
import { RequestJobCommandSchema } from "../../../../protocols/generated/typescript/job/v1/job_commands_pb.js";
import type { Job } from "../../../../protocols/generated/typescript/job/v1/job_pb.js";
import { JobSchema } from "../../../../protocols/generated/typescript/job/v1/job_pb.js";
import type { Operation } from "../../../../protocols/generated/typescript/job/v1/operation_pb.js";
import { OperationSchema } from "../../../../protocols/generated/typescript/job/v1/operation_pb.js";
import type { ClientCore } from "./core.js";
import { MindcladeError } from "./error.js";
import { listPage, type Page, withPageToken } from "./pagination.js";
import {
	commandContext,
	type ListOptions,
	prepareCall,
	type SdkCallOptions,
	type SubmitOptions,
} from "./request.js";
import { invokeUnary } from "./retry.js";

const REQUEST = "/mindclade.internal.job.v1.JobService/RequestJob";
const GET = "/mindclade.internal.job.v1.JobService/GetJob";
const LIST = "/mindclade.internal.job.v1.JobService/ListJobs";
const CANCEL = "/mindclade.internal.job.v1.JobService/CancelJob";
const DIGEST = /^sha256:[0-9a-f]{64}$/;

/** Durable-job facade over authoritative generated JobService types. */
export class Jobs {
	readonly #core: ClientCore;
	constructor(core: ClientCore) {
		this.#core = core;
	}

	async request(
		input: MessageInitShape<typeof RequestJobCommandSchema>,
		options: SubmitOptions,
	): Promise<readonly [Job, Operation]> {
		const command = clone(RequestJobCommandSchema, create(RequestJobCommandSchema, input));
		if (
			!leaf(command.jobKind) ||
			!artifact(command.configuration) ||
			(command.input !== undefined && !artifact(command.input)) ||
			(command.requestedJobId !== "" && !leaf(command.requestedJobId))
		) {
			throw MindcladeError.invalidArgument(
				"job request requires valid kind, optional ID, and content-addressed configuration",
			);
		}
		delete command.context;
		const call = prepareCall(this.#core.config, this.#core.runtime, options);
		command.context = digestContext(this.#core, call, options, RequestJobCommandSchema, command);
		const response = await invokeUnary(
			this.#core,
			call,
			REQUEST,
			options.idempotencyKey,
			(callOptions) =>
				this.#core.raw.jobs.requestJob(create(RequestJobRequestSchema, { command }), callOptions),
		);
		if (
			response.job === undefined ||
			response.operation === undefined ||
			!validJob(this.#core, response.job) ||
			!validOperation(this.#core, response.operation) ||
			response.job.operationId !== response.operation.operationId ||
			response.operation.jobId !== response.job.jobId
		) {
			throw MindcladeError.protocol("RequestJob response violated durable identity");
		}
		return [clone(JobSchema, response.job), clone(OperationSchema, response.operation)];
	}

	async get(name: string, ifNoneMatch = "", options: SdkCallOptions = {}): Promise<Job> {
		const canonical = resource(this.#core, name, "jobs");
		if (ifNoneMatch.length > 512 || /[\0\r\n]/.test(ifNoneMatch))
			throw MindcladeError.invalidArgument("job cache validator is invalid");
		const call = prepareCall(this.#core.config, this.#core.runtime, options);
		const response = await invokeUnary(
			this.#core,
			call,
			GET,
			undefined,
			(callOptions) =>
				this.#core.raw.jobs.getJob(
					create(GetJobRequestSchema, { name: canonical, ifNoneMatch }),
					callOptions,
				),
		);
		if (
			response.job === undefined ||
			response.job.jobId !== canonical ||
			!validJob(this.#core, response.job)
		)
			throw MindcladeError.protocol("GetJob response violated durable identity");
		return clone(JobSchema, response.job);
	}

	/** Returns the first page, which also iterates the whole cursor. */
	async list(
		input: MessageInitShape<typeof ListJobsRequestSchema> = {},
		options: ListOptions = {},
	): Promise<Page<Job, ListJobsResponse>> {
		const request = clone(ListJobsRequestSchema, create(ListJobsRequestSchema, input));
		const parent = project(this.#core);
		const pageSize = request.page?.pageSize ?? 0;
		if (
			(request.parent !== "" && request.parent !== parent) ||
			!Number.isInteger(pageSize) ||
			pageSize < 0 ||
			pageSize > 200 ||
			request.filter.trim() !== "" ||
			!["", "job_id"].includes(request.orderBy)
		)
			throw MindcladeError.invalidArgument("job list scope, page, filter, or order is invalid");
		request.parent = parent;
		return await listPage({
			cursor: (response) => response.page?.nextPageToken ?? "",
			fetch: async (pageToken) => {
				const paged = withPageToken(ListJobsRequestSchema, request, pageToken);
				const call = prepareCall(this.#core.config, this.#core.runtime, options);
				const response = await invokeUnary(
					this.#core,
					call,
					LIST,
					undefined,
					(callOptions) => this.#core.raw.jobs.listJobs(paged, callOptions),
				);
				if (response.jobs.some((value) => !validJob(this.#core, value)))
					throw MindcladeError.protocol("ListJobs response escaped configured scope");
				return { requestId: call.requestId, response: clone(ListJobsResponseSchema, response) };
			},
			items: (response) => response.jobs,
			limits: options.limits,
			pageSize,
			pageToken: request.page?.pageToken ?? "",
			signal: options.signal,
		});
	}

	async cancel(
		input: MessageInitShape<typeof CancelJobRequestSchema>,
		options: SubmitOptions,
	): Promise<Operation> {
		const request = clone(CancelJobRequestSchema, create(CancelJobRequestSchema, input));
		request.name = resource(this.#core, request.name, "jobs");
		if (
			request.etag.trim() === "" ||
			request.etag.length > 512 ||
			/[\0\r\n]/.test(request.etag) ||
			request.reason.length > 4096 ||
			request.reason.includes("\0")
		)
			throw MindcladeError.invalidArgument(
				"job cancellation requires a valid ETag and bounded reason",
			);
		delete request.context;
		const call = prepareCall(this.#core.config, this.#core.runtime, options);
		request.context = digestContext(this.#core, call, options, CancelJobRequestSchema, request);
		const response = await invokeUnary(
			this.#core,
			call,
			CANCEL,
			options.idempotencyKey,
			(callOptions) => this.#core.raw.jobs.cancelJob(request, callOptions),
		);
		if (
			response.operation === undefined ||
			response.operation.jobId !== request.name ||
			!validOperation(this.#core, response.operation)
		)
			throw MindcladeError.protocol("CancelJob response violated durable identity");
		return clone(OperationSchema, response.operation);
	}
}

export const project = (core: ClientCore): string =>
	`tenants/${core.config.identity.tenantId.replace(/^tenants\//, "")}/projects/${core.config.identity.projectId.replace(/^projects\//, "")}`;
export const resource = (core: ClientCore, input: string, collection: string): string => {
	const value = input.trim();
	if (leaf(value)) return `${collection}/${value}`;
	const compact = new RegExp(`^${collection}/([^/]+)$`).exec(value);
	if (compact !== null && leaf(compact[1] ?? "")) return value;
	const prefix = `${project(core)}/${collection}/`;
	if (value.startsWith(prefix) && leaf(value.slice(prefix.length)))
		return `${collection}/${value.slice(prefix.length)}`;
	throw MindcladeError.invalidArgument(`${collection} resource is outside configured scope`);
};
export const leaf = (value: string): boolean => /^[A-Za-z0-9_.-]{1,255}$/.test(value.trim());
const artifact = (value: { digest: string; mediaType: string } | undefined): boolean =>
	value !== undefined && DIGEST.test(value.digest) && value.mediaType.trim() !== "";
const validJob = (core: ClientCore, value: Job): boolean =>
	value.tenantId === core.config.identity.tenantId &&
	value.projectId === core.config.identity.projectId &&
	resourceId(value.jobId, "jobs") &&
	resourceId(value.operationId, "operations") &&
	value.resourceVersion > 0n &&
	value.state !== 0;
const validOperation = (core: ClientCore, value: Operation): boolean =>
	value.tenantId === core.config.identity.tenantId &&
	value.projectId === core.config.identity.projectId &&
	resourceId(value.operationId, "operations") &&
	value.resourceVersion > 0n &&
	value.state !== 0;
export const resourceId = (value: string, collection: string): boolean =>
	new RegExp(`^${collection}/[A-Za-z0-9_.-]{1,255}$`).test(value);
export const digestContext = <Schema extends DescMessage>(
	core: ClientCore,
	call: ReturnType<typeof prepareCall>,
	options: SubmitOptions,
	schema: Schema,
	request: MessageShape<Schema>,
) => {
	const context = commandContext(core.config, call, options);
	context.canonicalRequestDigest = `sha256:${createHash("sha256").update(toBinary(schema, request)).digest("hex")}`;
	return context;
};

File: internal/sdk/rust/src/jobs.rs
use std::sync::Arc;

use mindclade_protocols::{
    artifact::v1::ArtifactRef,
    internal::job::v1::{CancelJobRequest, GetJobRequest, ListJobsRequest, RequestJobRequest},
    job::v1::{Job, Operation, RequestJobCommand},
};
use prost::Message;
use sha2::{Digest, Sha256};

use crate::{
    CallOptions, ClientCore, Error, Page, Pages, SubmitOptions,
    request::{initial_page_token, page_request},
    retry::registered_method_policy,
};

const REQUEST: &str = "/mindclade.internal.job.v1.JobService/RequestJob";
const GET: &str = "/mindclade.internal.job.v1.JobService/GetJob";
const LIST: &str = "/mindclade.internal.job.v1.JobService/ListJobs";
const CANCEL: &str = "/mindclade.internal.job.v1.JobService/CancelJob";

/// Durable admitted-work helpers over authoritative generated `JobService` types.
#[derive(Clone)]
pub struct Jobs {
    core: Arc<ClientCore>,
}

impl Jobs {
    pub(crate) fn new(core: Arc<ClientCore>) -> Self {
        Self { core }
    }

    /// Admits a generated job command and returns its durable job and operation.
    ///
    /// # Errors
    ///
    /// Returns an error for malformed intent, invalid credentials, transport
    /// failure, or a response outside the configured identity scope.
    pub async fn request(
        &self,
        mut command: RequestJobCommand,
        options: SubmitOptions,
    ) -> Result<(Job, Operation), Error> {
        if !valid_leaf(&command.job_kind)
            || !valid_artifact(command.configuration.as_ref(), true)
            || !valid_artifact(command.input.as_ref(), false)
            || (!command.requested_job_id.is_empty() && !valid_leaf(&command.requested_job_id))
        {
            return Err(Error::invalid_argument(
                "job request requires a valid kind, optional ID, and content-addressed configuration",
            ));
        }
        command.context = None;
        let prepared = options.call.prepare(&self.core.config);
        let digest = protobuf_digest(&command);
        let mut context = prepared.command_context(&self.core.config, &options)?;
        context.canonical_request_digest = digest;
        command.context = Some(context);
        let key = options.idempotency_key.clone();
        let response = self
            .core
            .unary(
                RequestJobRequest {
                    command: Some(command),
                },
                &prepared,
                registered_method_policy(REQUEST),
                Some(&key),
                |transport, request| Box::pin(async move { transport.request_job(request).await }),
            )
            .await?
            .into_inner();
        let job = response
            .job
            .ok_or_else(|| Error::protocol("RequestJob response omitted its job"))?;
        let operation = response
            .operation
            .ok_or_else(|| Error::protocol("RequestJob response omitted its operation"))?;
        if !valid_job(&self.core, &job)
            || !valid_operation(&self.core, &operation)
            || job.operation_id != operation.operation_id
            || operation.job_id != job.job_id
        {
            return Err(Error::protocol(
                "RequestJob response changed durable identity",
            ));
        }
        Ok((job, operation))
    }

    /// Reads one durable job revision.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid scope or cache metadata, credentials,
    /// transport failure, or a malformed response.
    pub async fn get(
        &self,
        name: impl Into<String>,
        if_none_match: impl Into<String>,
        options: CallOptions,
    ) -> Result<Job, Error> {
        let name = canonical_resource(&self.core, &name.into(), "jobs")?;
        let if_none_match = if_none_match.into();
        if if_none_match.len() > 512 || contains_transport_control(&if_none_match) {
            return Err(Error::invalid_argument("job cache validator is invalid"));
        }
        let prepared = options.prepare(&self.core.config);
        let response = self
            .core
            .unary(
                GetJobRequest {
                    name: name.clone(),
                    if_none_match,
                },
                &prepared,
                registered_method_policy(GET),
                None,
                |transport, request| Box::pin(async move { transport.get_job(request).await }),
            )
            .await?
            .into_inner();
        let job = response
            .job
            .ok_or_else(|| Error::protocol("GetJob response omitted its job"))?;
        if job.job_id != name || !valid_job(&self.core, &job) {
            return Err(Error::protocol("GetJob response changed durable identity"));
        }
        Ok(job)
    }

    /// Lists one bounded project-scoped page and preserves opaque tokens.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid scope or query bounds, credentials,
    /// transport failure, or a response outside the configured project.
    pub fn list(
        &self,
        mut request: ListJobsRequest,
        options: CallOptions,
    ) -> Result<Pages<Job>, Error> {
        let parent = project_name(&self.core);
        if (!request.parent.is_empty() && request.parent != parent)
            || request
                .page
                .as_ref()
                .is_some_and(|page| page.page_size > 200)
            || !request.filter.trim().is_empty()
            || !matches!(request.order_by.as_str(), "" | "job_id")
        {
            return Err(Error::invalid_argument(
                "job list scope, page size, filter, or ordering is invalid",
            ));
        }
        request.parent = parent;
        let core = Arc::clone(&self.core);
        let token = initial_page_token(request.page.as_ref());
        Ok(Pages::new(
            move |page_token| {
                let core = Arc::clone(&core);
                let options = options.clone();
                let mut request = request.clone();
                async move {
                    request.page = Some(page_request(request.page.as_ref(), page_token));
                    let prepared = options.prepare(&core.config);
                    let response = core
                        .unary(
                            request,
                            &prepared,
                            registered_method_policy(LIST),
                            None,
                            |transport, request| {
                                Box::pin(async move { transport.list_jobs(request).await })
                            },
                        )
                        .await?;
                    let request_id = response.request_id().map(str::to_owned);
                    let response = response.into_inner();
                    if response.jobs.iter().any(|job| !valid_job(&core, job)) {
                        return Err(Error::protocol(
                            "ListJobs response escaped configured scope",
                        ));
                    }
                    Ok(Page::new(
                        response.jobs,
                        response.page,
                        response.read_time,
                        request_id,
                    ))
                }
            },
            token,
        ))
    }

    /// Records monotonic job cancellation under an `ETag` precondition.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid intent, credentials, transport failure,
    /// or a malformed operation response.
    pub async fn cancel(
        &self,
        mut request: CancelJobRequest,
        options: SubmitOptions,
    ) -> Result<Operation, Error> {
        request.name = canonical_resource(&self.core, &request.name, "jobs")?;
        if request.etag.trim().is_empty()
            || request.etag.len() > 512
            || contains_transport_control(&request.etag)
            || request.reason.len() > 4096
            || request.reason.contains('\0')
        {
            return Err(Error::invalid_argument(
                "job cancellation requires a valid ETag and bounded reason",
            ));
        }
        request.context = None;
        let prepared = options.call.prepare(&self.core.config);
        let digest = protobuf_digest(&request);
        let mut context = prepared.command_context(&self.core.config, &options)?;
        context.canonical_request_digest = digest;
        request.context = Some(context);
        let key = options.idempotency_key.clone();
        let expected_job = request.name.clone();
        let response = self
            .core
            .unary(
                request,
                &prepared,
                registered_method_policy(CANCEL),
                Some(&key),
                |transport, request| Box::pin(async move { transport.cancel_job(request).await }),
            )
            .await?
            .into_inner();
        let operation = response
            .operation
            .ok_or_else(|| Error::protocol("CancelJob response omitted its operation"))?;
        if operation.job_id != expected_job || !valid_operation(&self.core, &operation) {
            return Err(Error::protocol(
                "CancelJob response changed durable identity",
            ));
        }
        Ok(operation)
    }
}

pub(crate) fn project_name(core: &ClientCore) -> String {
    format!(
        "tenants/{}/projects/{}",
        core.config
            .identity
            .tenant_id()
            .trim_start_matches("tenants/"),
        core.config
            .identity
            .project_id()
            .trim_start_matches("projects/")
    )
}

pub(crate) fn canonical_resource(
    core: &ClientCore,
    value: &str,
    collection: &str,
) -> Result<String, Error> {
    let value = value.trim();
    if valid_leaf(value) {
        return Ok(format!("{collection}/{value}"));
    }
    if valid_compact_resource(value, collection) {
        return Ok(value.to_owned());
    }
    let prefix = format!("{}/{collection}/", project_name(core));
    if let Some(leaf) = value.strip_prefix(&prefix)
        && valid_leaf(leaf)
    {
        return Ok(format!("{collection}/{leaf}"));
    }
    Err(Error::invalid_argument(format!(
        "{collection} resource is outside configured project"
    )))
}

pub(crate) fn valid_leaf(value: &str) -> bool {
    let value = value.trim();
    !value.is_empty()
        && value.len() <= 255
        && value
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'_' | b'.'))
}

pub(crate) fn valid_compact_resource(value: &str, collection: &str) -> bool {
    value
        .strip_prefix(collection)
        .and_then(|rest| rest.strip_prefix('/'))
        .is_some_and(valid_leaf)
}

pub(crate) fn protobuf_digest(value: &impl Message) -> String {
    let bytes = value.encode_to_vec();
    format!("sha256:{:x}", Sha256::digest(bytes))
}

fn valid_artifact(value: Option<&ArtifactRef>, required: bool) -> bool {
    match value {
        None => !required,
        Some(value) => {
            valid_sha256(&value.digest)
                && !value.media_type.trim().is_empty()
                && value.size_bytes >= 0
        }
    }
}

fn valid_sha256(value: &str) -> bool {
    value.len() == 71
        && value.starts_with("sha256:")
        && value[7..]
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}

fn valid_job(core: &ClientCore, value: &Job) -> bool {
    value.tenant_id == core.config.identity.tenant_id()
        && value.project_id == core.config.identity.project_id()
        && valid_compact_resource(&value.job_id, "jobs")
        && valid_compact_resource(&value.operation_id, "operations")
        && value.resource_version > 0
        && value.state != 0
}

fn valid_operation(core: &ClientCore, value: &Operation) -> bool {
    value.tenant_id == core.config.identity.tenant_id()
        && value.project_id == core.config.identity.project_id()
        && valid_compact_resource(&value.operation_id, "operations")
        && value.resource_version > 0
        && value.state != 0
}

fn contains_transport_control(value: &str) -> bool {
    value.bytes().any(|byte| matches!(byte, 0 | b'\r' | b'\n'))
}
